### DISCUSSION TITLE
Plot multiple x/y arrays from a single message

### DISCUSSION THREAD

**Discussion by defunctzombie** [foxglove/community#176] [Q&A] [answered]
A common pattern in planning, perception, or controls is producing several candidate options (paths, trajectories, etc) at any time. These candidates may appear in a single message (as arrays of candidates). An engineer may want to visualize these different candidates against one another on a single plot.

The current plot panel allows a user to plot arbitrary x/y values from an array within a message using the current x-axis setting. This setting (like the accumulated) setting only allows the user to select 1 x-axis value and explicit multiple y-axis values.

We should provide a way to allow the user to select arrays of x/y values and plot each array of x/y values as a dataset.

**Comment by tonynajjar** [comment]
That would be extremely useful! I'm surprised there aren't more upvotes on this. I just wrote a similar issue https://github.com/orgs/foxglove/discussions/803 but this one formulates the feature request more generically

**Comment by mirsking** [comment]
+1

**Comment by kyoichi-sugahara** [comment]
@defunctzombie @tonynajjar 
I feel this feature is really necessary, and if currently no one has provided a way to implement this functionality, I'm thinking of trying to address it even in a hacky way using TypeScript. Do you have any information about this?

**Reply by defunctzombie** [reply]
I cannot think of an easy way to hack around this. It is a feature that does not exist in the plot panel - the ability to select different _x_ values for different series on a scatterplot.

**Comment by patonz** [comment]
still no update?
a workaround i found is to plot in chunks the static data, and then start with the realtime one. Works fine but im still facing the issue regarding the X where maybe those data are not normalized.

**Comment by lucafei** [comment]
+1

**Comment by xavier-guardian** [comment]
+1 

But this will probably need to be addressed more generally by allowing each series object to have its own x axis settings. 

Like these settings: 
![image](https://github.com/user-attachments/assets/82f473ca-cb28-47b8-a63b-f9865e7c6c49)

Value, Message Path, Range should be here:
![image](https://github.com/user-attachments/assets/61575773-2efa-42e7-95e9-5fcc4777f2d7)

So that for each series you can specify whether you want its timestamp to be it's x value, or whether you want another `message_path` to be the x value. That way you can have an arbitrary number of XY plots and the strangeness of "Path (accumulated)" could go away. In the old open source days I would open a PR :P

**Reply by indirectlylit** [reply]
Hello! Just wanted to update that we are actively working on adding the ability to specify independent x-values for each series.

> for each series you can specify whether you want its timestamp to be it's x value, or whether you want another message_path to be the x value

Initially we were planning to have the [X Axis value field](https://docs.foxglove.dev/docs/visualization/panels/plot#x-axis) remain constant but provide a way to specify the path independently for each series. I'd be interested to hear a bit more about the "mix and match" use cases I think you're describing here?

**Reply by xavier-guardian** [reply]
@indirectlylit 
> Initially we were planning to have the [X Axis value field](https://docs.foxglove.dev/docs/visualization/panels/plot#x-axis) remain constant but provide a way to specify the path independently for each series. I'd be interested to hear a bit more about the "mix and match" use cases I think you're describing here?

This is probably better than what I had in mind tbh. Thinking more clearly about it, they wouldn't really be able to share x axis max and min values since that defines the whole plot, so that would probably have to stay global, like y-axis. So your method just being able to specify what you want the x-coordinate of the "point pair (x,y)" to be in terms of a message path would make a lot of sense, imo. 

I wonder then if you ever care about the units of the axes? The y-axis is unit agnostic, so i suppose the x-axis could be too? Like if i plot a time series, and a x-y odometry plot together on the same plot, it would look stupid and be unreadable, BUT should be viable.

**Comment by indirectlylit** [comment, marked as answer]
Hi all! I'm happy to share that we recently shipped a feature which allows you to set X values independently for each series in path-based (accumulated and current) XY plots.

It's available on web now, and will be in the next desktop release. If you try it and have any feedback, please let us know!

**Reply by xavier-guardian** [reply]
Thank you for the responsiveness!

**Reply by xavier-guardian** [reply]
@indirectlylit wooo! This is exactly what I wanted. Thank you! Will spread the good word of Foxglove. 

![image](https://github.com/user-attachments/assets/7bc22c2d-7550-406e-b1ec-d030d641b070)
